### ISSUE TITLE
2021/09/29 - 2021/10/07

### ISSUE THREAD

**Issue by njs03332** [njs03332/nlp_study#3]
3장 3.6.1 (The Yelp Review Dataset) 까지 공부해오기
* 정리: 한단비님
* 보충 정리: 김유리님

1장 1.5 연습문제 풀어오기

**Comment by danbi5228** [comment]
\>>>>> 정리하다보니 양이 많네요.. 아직 진행중입니다 ㅠ 죄송합니다

## 3장. 신경만의 기본 구성 요소
### 3.1 퍼셉트론: 가장 간단한 신경망
* 퍼셉트론 perceptron
   * 가장 간단한 신경망
   * 유닛 unit 이 하나 있는 신경망
   * 복잡한 신경망의 구성 요소
* 퍼셉트론 유닛
   * 입력 x, 출력 y, 3개의 다이얼로 구성됨
      * 3개의 다이얼: 가중치 w (weight), 절편 b (bias), 활성화 함수 f (function)
      * 가중치, 절편 : 데이터에서 학습
      * 활성화 함수: 신경망에서 데이터의 복잡한 관계를 감지하는데 사용
         * 신경망과 타깃 출력을 기반으로 설계자의 직관에 따라 결정됨
         * 수학적으로 표현한다면 -> y = f(wx + b)
         * 선형함수와 비선형 함수의 조합
         * 아핀 변환 affine transform : 선형 함수 표현인 wx + b 를 지칭함

### 3.2 활성화 함수
* 널리 사용하는 비선형 함수 4가지 - 시그모이드, 하이퍼볼릭 탄젠트, 렐루, 소프트맥스

#### 3.2.1 시그모이드 sigmoid
* 신경망 분야의 초창기부터 사용한 활성화 함수
* 임의의 실숫값을 받아 0과 1 사이의 범위로 압축
* 식: f(x) = 1 / (1+ (e^-x) )
* 미분할 수 있는 부드러운 함수
* pytorch : torch.sigmid()
* 특징
   * 입력 범위 대부분에서 매우 빠르게 포화되어, 아래 문제를 야기함 <<<<< ???
      * 그레이디언트 소실 문제 vanishing gradient problem
         * 그레이디언트가 0이 되는 문제
      * 그레이디언트 폭주 문제 exploding gradient problem
         * 그레이디언트가 발산하여 부동소수 오버플로가 발생하는 문제
   * 그래서 신경망에서 시그모이드 활성화 함수는 거의 출력층에서만 사용 - **출력을 확률로 압축**하는데 사용

#### 3.2.2 하이퍼볼릭 탄젠트 hyperbolic tangent (tanh)
* 시그모이드 함수의 변종 (시그모이드 함수의 선형 변환. 책 65페이지 참고)
* 식: f(x) = tanh x = (e^x - e^-x) / (e^x + e^-x)
* 시그모이드처럼 하이버폴릭 탄젠트도 압축함수
* 시그모이드와 달리 실숫값의 범위를 (-무한, +무한) 에서 (-1, +1)로 바꿈
* pytorch : torch.tanh()

#### 3.2.3 렐루 rectified linear unit (ReLU)
* 최근 딥러닝 혁신과 연관된, 가장 중요한 활성화 함수
* 식: f(x) = max(0, x)
* 음숫값을 0으로 자름
* pytorch : torch.nn.ReLU()
* 특징
   * 음수를 제거해 그레디언트 소실문제에 도움이 됨
   * 그러나 '죽은 렐루' 문제가 발생할 수 있음
      * 시간이 지나서 신경망의 특정 출력이 0이 되면 다시 돌아오지 않는 문제
   * '죽을 렐루' 문제를 줄이기 위해 등장한 변종: 리키 렐루 (LeakyReLU), PReLU (parametric ReLU)
      * 식: f(x) = max(x, ax) (변종 둘다 동일한 식)
      * PReLU : 누수 파라미터 a는 학습되는 파라미터
      * 리키 렐루 : 누수 파라미터 a는 사전에 설정하는 하이퍼파라미터

#### 3.2.4 소프트맥스 softmax
* 시그모이드 함수처럼 신경망 유닛의 출력을 0과 1 사이로 압축
* 모든 출력의 합으로 각 출력을 나누어 k개 클래스에 대한 이산 확률 분포를 만듦
* 식: p.67 참고
* pytorch : torch.nn.Softmax()
* 특징
   * 출력의 합이 1 - 분류 작업의 출력을 해석하는데 유용 (이진 분류: 시그모이드, 다중분류: 소프트맥스)
   * 보통 확률 기반의 목적 함수인 범주형 크로스 엔트로피(3.3.2) 와 함께 사용

### 3.3 손실 함수
* 손실 함수: 정답과 예측을 입력으로 받아 실숫값 점수를 만듦. 이 점수가 높을 수록 모델의 예측 성능은 나빠짐
* 가장 널리 사용하는 손실함수 3가지 - 평균 제곱 오차 손실, 범주형 크로스 엔트로피 손실, 이진 크로스 엔트로피 손실

#### 3.3.1 평균 제곱 오차 손실 mean squared error (MSE)
* 신경망 출력과 타깃이 연속값인 회귀 문제에서 널리 사용
* 식: p.69 참고
* 예측과 타깃값의 차이를 제곱하여 평균한 값
* 회귀 문제에서 사용할 수 있는 다른 손실 함수
   * 평균 절댓값 오차 mean absolute error (MAE)
   * 평균 제곱근 오차 root mean squared error (RMSE)
   * 둘다 출력과 타깃 사이의 거리를 실숫값으로 계산
* pytorch: torch.nn.MSELoss()

#### 3.3.2 범주형 크로스 엔트로피 손실 (categorical cross-entropy 손실)
* 일반적으로 출력을 클래스 소속 확률에 대한 예측으로 이해할 수 있는 다중 분류 문제에서 사용
* 타깃: 모든 클래스에 대한 다항 분포를 나타내는 원소 n개로 이루어진 벡터
   * 하나의 클래스만 정답이면 이 벡터는 원-핫 벡터
* 출력: 원수 n개로 구성된 벡터. 다항 분포에 대한 신경망의 예측을 나타냄
* 식: p.69 참고
* 타깃 벡터와 출력 벡터를 비교하여 손실 계산
* 정답 클래스의 확률은 1에 가깝고 다른 클래스의 확률은 0에 가까운 상태가 바람직함
* pytorch: torch.nn.CrossEntropyLoss()

\+ 신경망 출력과 손실함수 간의 미묘한 관계를 결정하는 정보 4가지 (코드 작성시 유의할 점?)
   \- 1. 수의 범위 제한
   \- 2. 소프트맥스 함수에 사용한 지수함수의 입력이 음수이면 기하급수적으로 작은 숫자가 됨. 양수이면 기하급수적으로 큰 숫자가 됨
   \- 3. 신경망의 출력은 소프트맥스 함수를 적용하기 직전의 벡터로 가정
   \- 4. 로그함수는 지수함수의 역함수이고 log(exp(x)) 는 x와 동일 (로그함수 밑이 파이토치 로그처럼 자연상수 e일때 해당)
-> 이 4가지 정보를 기반으로 소프트맥스의 지수함수와 크로스 엔트로피의 로그함수를 간소화해서 너무 작거나 큰 값을 피함

#### 3.3.3 이진 크로스 엔트로피 손실
* 클래스 두 개를 구별하는 이진분류 문제에서 유용. 범주형 크로스 엔트로피 함수는 다중 클래스 분류 문제에서 유용
* pytorch: torch.nn.BCELoss()

### 3.4 지도 학습 훈련 알아보기
* 결정 경계 (decision boundary) (= 초평면 hyperplane)
   * 2차원 데이터 포인트를 클래스 두 개중 하나로 분류하고자 한다면, 직선 하나를 학습하겠다는 의미.
   * 여기서 이 직선을, 한 클래스 포인트를 다른 클래스와 구별하는 결정경계 또는 초평면 이라고 함

#### 3.4.1 예제 데이터 만들기
* get_toy_data() 함수 내용 확인
#### 3.4.2 모델 선택
* 퍼셉트론 사용. 퍼셉트론의 활성화 함수는 시그모이드
* 퍼셉트론의 출력은 데이터 포인트가 클래스 1일 확률
#### 3.4.3 확률을 클래스로 변환하기
* 이진 분류 문제 이므로 결정경계를 적용해 출력 확률을 두 개의 클래스로 바꾸어야 함. 1 또는 0
* 일반적으로 결정경계는 0.5. 하지만 실전에서 만족스러운 분류 정밀도 precision을 얻으려면 검증데이터 셋으로 이 하이퍼파라미터를 튜닝해야함
#### 3.4.4 손실 함수 선택
* 모델의 출력이 확률이라면 가장 적절한 손실함수는 크로스 엔트로피 기반의 손실 함수
* 이진 출력을 만들므로 이진 크로스 엔트로피 손실 함수 사용
#### 3.4.5 옵티마이저 선택
* 옵티마이저: 손실함수가 예측과 타깃사이에서 측정한 오차 신호를 가지고 모델의 가중치를 업데이트
* 학습률 learning ratio: 오차 신호가 가중치 업데이트에 얼마나 미치는지 조절하는 하이퍼 파라미터
   * 다양한 학습률을 시도하여 결과를 비교해야 함
   * 큰 학습률은 가중치를 크게 바꾸고 수렴에 영향을 미칠 수 있음
   * 작은 학습률은 훈련 진행 속도가 심하게 느려질 수 있음
* 현재 선호하는 옵티마이저
   * Adagrad
   * Adam - 예시에서 선택
   * 둘 다 적응형 옵티마이저로, 시간에 따라 업데이트 정보를 사용 
   * 항상 여러 옵티마이저를 시도해봐야 하고, 학습률은 기본값을 가장 먼저 사용하는 것을 권장
#### 3.4.6 모두 합치기: 그레이디언트 기반의 지도 학습
* 각 파라미터의 그레이디언트는 이 파라미터에 대한 손실값의 순간 변화율(=기울기)을 의미함. 각 파라미터가 손실함수에 얼마나 많이 기여하는지 알 수 있다는 의미
* 에포크 : 완전한 훈련 반복 한 번을 의미
* 에포크 당 배치 개수가 데이터 셋의 배치 개수와 같다면, 에포크가 데이터셋에 대한 완전한 반복 한 번이 됨
* 지정된 에포크 횟수나 다른 종료조건을 만족할 때까지 훈련하게 됨 (훈련 중지시점: 3.5.3)

### 3.5 부가적인 훈련 개념

---

* 참고할만한 사이트: PyTorch로 시작하는 딥 러닝 입문 https://wikidocs.net/book/2788

**Comment by njs03332** [comment]
## Chapter 3
### 3.3 손실함수
#### 3.3.2 범주형 크로스 엔트로피 손실
- `CrossEntropyLoss()` 클래스를 사용하면, 소프트맥스 함수를 사용하지 않고 확률 분포를 최적화할 수 있음
  - 입력값이 클수록 기하급수적으로 값이 커지는 지수 함수의 특징을 보완하기 위해, 소프트맥스 함수의 분모, 분자에서 가장 큰 출력값을 빼준다.
  - 여기에 로그를 취해 로그 확률을 계산한다: `LogSoftmax()`
  - 로그 확률을 이용해 크로스 엔트로피 손실을 계산한다: `NLLLoss()`
  - `CrossEntropyLoss()`는 이 두 클래스를 감싸서 편리하게 사용할 수 있도록 해줌
